bson/src/main/org/bson/codecs/pojo/ConventionUseGettersAsSettersImpl.java

```diff
@@ -0,0 +1,115 @@
+        for (PropertyModelBuilder<?> propertyModelBuilder : classModelBuilder.getPropertyModelBuilders()) {
+            if (!(propertyModelBuilder.getPropertyAccessor() instanceof PropertyAccessorImpl)) {
+                throw new CodecConfigurationException(format("The USE_GETTER_AS_SETTER_CONVENTION is not compatible with "
+                        + "propertyModelBuilder instance that have custom implementations of org.bson.codecs.pojo.PropertyAccessor: %s",
+                        propertyModelBuilder.getPropertyAccessor().getClass().getName()));
+            }
+            PropertyAccessorImpl<?> defaultAccessor = (PropertyAccessorImpl<?>) propertyModelBuilder.getPropertyAccessor();
+            PropertyMetadata<?> propertyMetaData = defaultAccessor.getPropertyMetadata();
+            if (!propertyMetaData.isDeserializable() && propertyMetaData.isSerializable()
+                    && isMapOrCollection(propertyMetaData.getTypeData().getType())) {
+                setPropertyAccessor(propertyModelBuilder);
+            }
+        }
+    }
+
+    private <T> boolean isMapOrCollection(final Class<T> clazz) {
+        return Collection.class.isAssignableFrom(clazz) || Map.class.isAssignableFrom(clazz);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> void setPropertyAccessor(final PropertyModelBuilder<T> propertyModelBuilder) {
+        propertyModelBuilder.propertyAccessor(new PrivateProperyAccessor<T>(
+                (PropertyAccessorImpl<T>) propertyModelBuilder.getPropertyAccessor()));
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    private static final class PrivateProperyAccessor<T> implements PropertyAccessor<T> {
+        private final PropertyAccessorImpl<T> wrapped;
+
+        private PrivateProperyAccessor(final PropertyAccessorImpl<T> wrapped) {
+            this.wrapped = wrapped;
+        }
+
+        @Override
+        public <S> T get(final S instance) {
+            return wrapped.get(instance);
+        }
+
+        @Override
+        public <S> void set(final S instance, final T value) {
+            if (value instanceof Collection) {
+                mutateCollection(instance, (Collection) value);
+            } else if (value instanceof Map) {
+                mutateMap(instance, (Map) value);
+            } else {
+                setError(format("Unexpected type: '%s'", value.getClass()), null);
+            }
+        }
+
+        private <S> void mutateCollection(final S instance, final Collection value) {
+            T originalCollection = get(instance);
+            Collection<?> collection = ((Collection<?>) originalCollection);
+            if (collection == null) {
+                setError("The getter returned null.", null);
+            } if (!collection.isEmpty()) {
+                setError("The getter returned a non empty collection.", null);
+            } else {
+                try {
+                    collection.addAll(value);
```

Comment:
It seems like it would be possible to violate the type system here since we're using raw types.  I'm thinking of a case where a property started as something like

     List<Foo> getBazList();

and was used to encode some documents.  Then it was changed to

     List<Bar> getBazList();

would this end up silently adding Foo instances to the collection?  Is there a way to prevent that by keeping track of the expected generic type of the collection and throwing an exception if "instance" type is not assignable to that generic type?

Reply:
I gather by the test that it's prevented already.